File: kiko-api/src/services/judge/tokenIntelligenceLayer.ts
/**
 * Token Intelligence Layer
 * Aggregates all 5 intelligence sub-modules
 */

import {
    TokenIntelligenceLayer,
    TokenIntelligenceLabel,
    ProjectIdentity,
    SocialPresence,
    NarrativeStrength,
    WebsiteQuality,
    RiskSignals,
} from '../../types/judgeTypes.js';

export interface TokenIntelligenceInput {
    projectIdentity: ProjectIdentity;
    socialPresence: SocialPresence;
    narrativeStrength: NarrativeStrength;
    websiteQuality: WebsiteQuality;
    riskSignals: RiskSignals;
}

/**
 * Aggregate all Token Intelligence modules into final layer
 */
export function evaluateTokenIntelligenceLayer(
    input: TokenIntelligenceInput
): TokenIntelligenceLayer {
    const reasons: string[] = [];
    const riskTags: string[] = [];

    // Calculate weighted score
    // Weights: Risk (35%), Social (25%), Identity (20%), Narrative (10%), Website (10%)
    const score = (
        input.riskSignals.score * 0.35 +
        input.socialPresence.score * 0.25 +
        input.projectIdentity.score * 0.20 +
        input.narrativeStrength.score * 0.10 +
        input.websiteQuality.score * 0.10
    );

    // Determine label
    let label: TokenIntelligenceLabel;
    if (score >= 0.75) {
        label = 'strong';
        reasons.push('项目信息完整，真实性高');
    } else if (score >= 0.50) {
        label = 'medium';
        reasons.push('项目信息中等，存在一定风险');
    } else if (score >= 0.25) {
        label = 'weak';
        reasons.push('项目信息不足，真实性存疑');
    } else {
        label = 'danger';
        reasons.push('⚠️ 项目信息严重不足或有明显危险信号');
    }

    // Generate risk tags based on individual scores

    // Risk Signals tags
    if (input.riskSignals.scam_reports_found) {
        riskTags.push('scam-reports');
    }
    if (input.riskSignals.rug_reports_found) {
        riskTags.push('rug-reports');
    }
    if (input.riskSignals.honeypot_reports_found) {
        riskTags.push('honeypot');
    }

    // Social tags
    if (input.socialPresence.score >= 0.7) {
        riskTags.push('strong-social');
        reasons.push('社交媒体活跃度高');
    } else if (input.socialPresence.score < 0.3) {
        riskTags.push('weak-social');
        reasons.push('社交媒体活跃度低或无活动');
    }

    if (input.socialPresence.has_kol_mentions) {
        riskTags.push('kol-mentioned');
    }

    // Identity tags
    if (input.projectIdentity.score < 0.3) {
        riskTags.push('no-identity');
    }
    if (!input.projectIdentity.contract_verified) {
        riskTags.push('unverified-contract');
    }

    // Narrative tags
    if (input.narrativeStrength.narrative_type !== 'unknown') {
        riskTags.push(input.narrativeStrength.narrative_type);
    }
    if (input.narrativeStrength.score < 0.4) {
        riskTags.push('weak-narrative');
    }

    // Website tags
    if (!input.websiteQuality.has_website) {
        riskTags.push('no-website');
    } else if (!input.websiteQuality.has_docs) {
        riskTags.push('no-docs');
    }
    if (input.websiteQuality.has_product) {
        riskTags.push('has-product');
    }

    // Age tags
    if (input.projectIdentity.contract_age_hours < 1) {
        riskTags.push('very-early');
    } else if (input.projectIdentity.contract_age_hours < 24) {
        riskTags.push('early-stage');
    }

    return {
        project_identity: input.projectIdentity,
        social_presence: input.socialPresence,
        narrative_strength: input.narrativeStrength,
        website_quality: input.websiteQuality,
        risk_signals: input.riskSignals,
        token_intelligence_score: score,
        token_intelligence_label: label,
        risk_tags: riskTags,
        reasons,
    };
}

File: kikoapi/internal/services/judge/token_intelligence_layer.go
package judge

import (
	"kikoapi/internal/types"
)

// TokenIntelligenceInput aggregates the 5 sub-module outputs for token intelligence.
type TokenIntelligenceInput struct {
	ProjectIdentity   types.ProjectIdentity
	SocialPresence    types.SocialPresence
	NarrativeStrength types.NarrativeStrength
	WebsiteQuality    types.WebsiteQuality
	RiskSignals       types.RiskSignals
}

// EvaluateTokenIntelligenceLayer aggregates sub-modules into TokenIntelligenceLayer.
func EvaluateTokenIntelligenceLayer(input TokenIntelligenceInput) types.TokenIntelligenceLayer {
	var reasons, riskTags []string

	score := input.RiskSignals.Score*0.35 +
		input.SocialPresence.Score*0.25 +
		input.ProjectIdentity.Score*0.20 +
		input.NarrativeStrength.Score*0.10 +
		input.WebsiteQuality.Score*0.10

	var label types.TokenIntelligenceLabel
	switch {
	case score >= 0.75:
		label = types.TokenIntelligenceStrong
		reasons = append(reasons, "Project information is complete and credible")
	case score >= 0.50:
		label = types.TokenIntelligenceMedium
		reasons = append(reasons, "Project information is moderate with some risk")
	case score >= 0.25:
		label = types.TokenIntelligenceWeak
		reasons = append(reasons, "Project information is limited; credibility is questionable")
	default:
		label = types.TokenIntelligenceDanger
		reasons = append(reasons, "Project information is severely lacking or has clear red flags")
	}

	if input.RiskSignals.ScamReportsFound {
		riskTags = append(riskTags, "scam-reports")
	}
	if input.RiskSignals.RugReportsFound {
		riskTags = append(riskTags, "rug-reports")
	}
	if input.RiskSignals.HoneypotReportsFound {
		riskTags = append(riskTags, "honeypot")
	}
	if input.SocialPresence.Score >= 0.7 {
		riskTags = append(riskTags, "strong-social")
		reasons = append(reasons, "Social presence is strong")
	} else if input.SocialPresence.Score < 0.3 {
		riskTags = append(riskTags, "weak-social")
		reasons = append(reasons, "Social presence is weak or inactive")
	}
	if input.SocialPresence.HasKolMentions {
		riskTags = append(riskTags, "kol-mentioned")
	}
	if input.ProjectIdentity.Score < 0.3 {
		riskTags = append(riskTags, "no-identity")
	}
	if !input.ProjectIdentity.ContractVerified {
		riskTags = append(riskTags, "unverified-contract")
	}
	if input.NarrativeStrength.NarrativeType != "unknown" {
		riskTags = append(riskTags, input.NarrativeStrength.NarrativeType)
	}
	if input.NarrativeStrength.Score < 0.4 {
		riskTags = append(riskTags, "weak-narrative")
	}

	return types.TokenIntelligenceLayer{
		ProjectIdentity:        input.ProjectIdentity,
		SocialPresence:         input.SocialPresence,
		NarrativeStrength:      input.NarrativeStrength,
		WebsiteQuality:         input.WebsiteQuality,
		RiskSignals:            input.RiskSignals,
		TokenIntelligenceScore: score,
		TokenIntelligenceLabel: label,
		RiskTags:               riskTags,
		Reasons:                reasons,
	}
}

// DefaultTokenIntelligenceInput returns a neutral input when no external data is available.
func DefaultTokenIntelligenceInput() TokenIntelligenceInput {
	return TokenIntelligenceInput{
		ProjectIdentity:   types.ProjectIdentity{Score: 0.5},
		SocialPresence:    types.SocialPresence{Score: 0.5},
		NarrativeStrength: types.NarrativeStrength{Score: 0.5, NarrativeType: "unknown"},
		WebsiteQuality:    types.WebsiteQuality{Score: 0.5},
		RiskSignals:       types.RiskSignals{Score: 0.5},
	}
}
